pkg/gatewayconfigurationserver/gcsv2/get_gateway.go

```diff
@@ -0,0 +1,253 @@
+)
+
+type antennaLocation struct {
+	Latitude  float64 `json:"latitude,omitempty"`
+	Longitude float64 `json:"longitude,omitempty"`
+	Altitude  int32   `json:"altitude,omitempty"`
+}
+
+type oauth2Token struct {
+	AccessToken string `json:"access_token"`
+	ExpiresIn   uint32 `json:"expires_in"`
+}
+
+type attributes struct {
+	Description string `json:"description"`
+}
+
+type router struct {
+	ID          string `json:"id,omitempty"`
+	MQTTAddress string `json:"mqtt_address"`
+}
+
+type gatewayInfoResponse struct {
+	ID               string           `json:"id,omitempty"`
+	Attributes       *attributes      `json:"attributes,omitempty"`
+	FrequencyPlan    string           `json:"frequency_plan"`
+	FrequencyPlanURL string           `json:"frequency_plan_url"`
+	AutoUpdate       bool             `json:"auto_update"`
+	FirmwareURL      string           `json:"firmware_url,omitempty"`
+	AntennaLocation  *antennaLocation `json:"antenna_location,omitempty"`
+	Token            *oauth2Token     `json:"token,omitempty"`
+	Router           *router          `json:"router,omitempty"`
+	FallbackRouters  []*router        `json:"fallback_routers,omitempty"`
+}
+
+var errUnauthenticated = errors.DefineUnauthenticated("unauthenticated", "call was not authenticated")
+
+func (s *Server) handleGetGateway(c echo.Context) error {
+	ctx := c.Request().Context()
+
+	gatewayIDs := ttnpb.GatewayIdentifiers{
+		GatewayID: c.Param("gateway_id"),
+	}
+	if err := gatewayIDs.ValidateContext(ctx); err != nil {
+		return err
+	}
+
+	registry, err := s.getRegistry(ctx, &gatewayIDs)
+	if err != nil {
+		return err
+	}
+	gateway, err := registry.Get(ctx, &ttnpb.GetGatewayRequest{
+		GatewayIdentifiers: gatewayIDs,
+		FieldMask: types.FieldMask{Paths: []string{
+			"description",
+			"attributes",
+			"frequency_plan_id",
+			"auto_update",
+			"update_channel",
+			"antennas",
+			"gateway_server_address",
+		}},
+	}, s.getAuth(ctx))
+	if err != nil {
+		return err
+	}
+
+	md := rpcmetadata.FromIncomingContext(ctx)
+	switch md.AuthType {
+	case "bearer":
+		if err := rights.RequireGateway(ctx, gatewayIDs, ttnpb.RIGHT_GATEWAY_INFO); err != nil {
+			return err
+		}
+	case "key":
+		keyAttr, ok := gateway.Attributes["key"]
+		if !ok || md.AuthValue != keyAttr {
+			return errUnauthenticated
+		}
+	default:
+		gateway = gateway.PublicSafe()
+	}
+
+	freqPlanURL := s.inferServerAddress(c) + compatAPIPrefix + "/frequency-plans/" + gateway.FrequencyPlanID
+
+	var rtr *router
+	if gateway.GatewayServerAddress != "" {
+		rtr = &router{
+			ID: gateway.GatewayServerAddress,
+		}
+		rtr.MQTTAddress, err = s.inferMQTTAddress(gateway.GatewayServerAddress)
+		if err != nil {
+			return err
+		}
+	}
+
+	response := &gatewayInfoResponse{
+		ID:               gateway.GatewayID,
+		FrequencyPlan:    gateway.FrequencyPlanID,
+		FrequencyPlanURL: freqPlanURL,
+		AutoUpdate:       gateway.AutoUpdate,
+	}
+
+	if rtr != nil {
+		response.Router = rtr
+		response.FallbackRouters = []*router{rtr}
```

Comment:
This can be left empty right?

Reply:
If we leave it empty the gRPC forwarder will fetch all routers from the discovery server, and we only want the one we specify.